enterprise/server/backends/redis_execution_collector/redis_execution_collector.go

```diff
@@ -150,3 +264,42 @@ func (c *collector) DeleteExecutions(ctx context.Context, iid string) error {
 func (c *collector) DeleteInvocationLinks(ctx context.Context, executionID string) error {
 	return c.rdb.Del(ctx, getInvocationLinkKey(executionID)).Err()
 }
+
+func (c *collector) DeleteReverseInvocationLinks(ctx context.Context, invocationID string) error {
+	return c.rdb.Del(ctx, getReverseInvocationLinkKey(invocationID)).Err()
+}
+
+func unmarshalStoredInvocationLinks(serializedResults []string) ([]*sipb.StoredInvocationLink, error) {
+	res := make([]*sipb.StoredInvocationLink, 0)
+	for _, serializedResult := range serializedResults {
+		link := &sipb.StoredInvocationLink{}
+		if err := proto.Unmarshal([]byte(serializedResult), link); err != nil {
+			return nil, err
+		}
+		res = append(res, link)
+	}
+	return res, nil
+}
+
+func mergeExecutionUpdates(serializedResults []string) (*repb.StoredExecution, error) {
+	out := &repb.StoredExecution{}
+	for _, serializedResult := range serializedResults {
+		event := &repb.StoredExecution{}
+		if err := proto.Unmarshal([]byte(serializedResult), event); err != nil {
+			return nil, err
+		}
+		// Copy fields from the latest event to the output proto. proto.Merge()
+		// is a convenient way to do this. Note that this will concatenate
+		// repeated fields, which is fine for now since repeated fields are not
+		// currently populated as execution events.
```

Comment:
lol this is a scary comment if I've ever read one!

Will this go away once we're fully on ch? Is it only ever merging 2 things? If either of those things are not true, you might think about how we can detect if this ever becomes a problem.

Reply:
This is really just taking the place of the mysql upserts that we are doing today, and mysql doesn't support repeated fields, so I didn't think this would be an issue. If we do eventually support storing in-progress executions in ClickHouse, then this might be a problem. I'll add a little test that makes repeated fields are handled if we introduce new ones.

Reply:
(Done)